File: built-in plugins dsp/mb-fm-bell.cpp
/**
 * MB FM Bell
 * Category : instrument
 * Type     : fm
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Crystalline FM bells
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_FM_BELL_H
#define MB_FM_BELL_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbFmBell : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-fm-bell";
    static constexpr const char* PLUGIN_NAME    = "MB FM Bell";
    static constexpr const char* PLUGIN_TYPE    = "fm";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float ratio = 3.5f;  // range [1, 8]
    float modIndex = 5f;  // range [0, 20]
    float volume = 0.7f;  // range [0, 1]
    };

    MbFmBell() = default;
    ~MbFmBell() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.ratio = std::clamp(params.ratio, 1f, 8f);
        params.modIndex = std::clamp(params.modIndex, 0f, 20f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB FM Bell
        return input;
    }
};

#endif // MB_FM_BELL_H

File: built-in plugins dsp/mb-fm-bell.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbFmBellPlugin: PluginDefinition = { id: 'mb-fm-bell', slug: 'mb-fm-bell', name: 'MB FM Bell', category: 'instrument', type: 'fm', version: '1.0.0', description: 'Crystalline FM bells', author: 'Max Booster', grade: 'A', oscillators: [{ type: 'sine', detune: 0, gain: 0.7 }], envelope: { attack: 0.001, decay: 1.5, sustain: 0.2, release: 1.0 }, parameters: [{ id: 'ratio', name: 'Ratio', type: 'float', defaultValue: 3.5, minValue: 1, maxValue: 8, automatable: true }, { id: 'modIndex', name: 'Mod Index', type: 'float', defaultValue: 5, minValue: 0, maxValue: 20, automatable: true }, { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.7, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { ratio: 3.5, modIndex: 5, volume: 0.7 } };

export default MbFmBellPlugin;
